### PR TITLE
Add DataTransfer interface

### PR DESCRIPTION
**Pull request by adrianheine**
According to https://html.spec.whatwg.org/multipage/interaction.html#the-datatransfer-interface.

### CODE REVIEW

File: lib/dom.js [hunk 1]
```diff
@@ -53,6 +53,40 @@ declare class FileList {
...
+declare class DataTransfer {
+  dropEffect: string;
+  effectAllowed: string;
+
+  items: DataTransferItemList; // readonly
+
+  setDragImage(image: Element, x: number, y: number);
+
+  /* old interface */
+  types: array<string>; // readonly
```
Comment (gabelevi) [inline lib/dom.js]: `Array<string>`

File: lib/dom.js [hunk 2]
```diff
@@ -53,6 +53,40 @@ declare class FileList {
...
+  getData(format: string): string;
+  setData(format: string, data: string);
+  clearData(format: ?string);
```
Comment (gabelevi) [inline lib/dom.js]: With optional things I've been doing `clearData(format?: string);`

File: lib/dom.js [hunk 3]
```diff
@@ -53,6 +53,40 @@ declare class FileList {
...
+  add(data: string, type: string): ?DataTransferItem;
+  add(data: File): ?DataTransferItem;
+  remove(index: number);
+  clear();
+};
+
+declare class DataTransferItem {
+  kind: string; // readonly
+  type: string; // readonly
+  getAsString(_callback: ?(data: string) => void);
```
Comment (gabelevi) [inline lib/dom.js]: I've been typing the return type of callbacks as `mixed` recently. People really like to use arrow functions as callbacks, like

```JavaScript
  item.getAsString(data => foo(data))
```

and if `foo()` doesn't return `void` then you would get a type error.

**Comment by gabelevi** [comment]
Looks great! I have a few nits, but I can address them when as I merge this in. I'll get it imported, reviewed and merged ASAP. Thanks for the contribution!

File: lib/dom.js [hunk 4]
```diff
@@ -53,6 +53,40 @@ declare class FileList {
...
 
+/* DataTransfer */
+
+declare class DataTransfer {
+  dropEffect: string;
+  effectAllowed: string;
+
+  items: DataTransferItemList; // readonly
+
+  setDragImage(image: Element, x: number, y: number);
```
Comment (gabelevi) [inline lib/dom.js]: Oh, and return types are required for methods in `declare class`. So this kind of line should be

```JavaScript
setDragImage(image: Element, x: number, y: number): void;
```

or it won't parse

**Comment by adrianheine** [comment]
Adressed some of the issues.

File: lib/dom.js [hunk 5]
```diff
@@ -53,6 +53,40 @@ declare class FileList {
...
+  files: FileList; // readonly
+}
+
+declare class DataTransferItemList {
+  length: number; // readonly
+  [index: number]: DataTransferItem;
+  add(data: string, type: string): ?DataTransferItem;
+  add(data: File): ?DataTransferItem;
+  remove(index: number): void;
+  clear();
```
Comment (mroch) [inline lib/dom.js]: here's another that's missing a return type. you can run the tests to double check, `./runtests.sh bin/flow dom`

File: lib/dom.js [hunk 6]
```diff
@@ -53,6 +53,40 @@ declare class FileList {
...
+};
```
Comment (mroch) [inline lib/dom.js]: can you remove these `;`s? thanks!

File: lib/dom.js (hunk 5 above)
Comment (adrianheine) [inline lib/dom.js]: They fail for me for unrelated reasons.

**Pull request closed without merging**